substrate/frame/identity/src/lib.rs

```diff
@@ -886,6 +966,181 @@ pub mod pallet {
 			});
 			Ok(())
 		}
+
+		/// Add an `AccountId` with permission to grant usernames with a given `suffix` appended.
+		/// The authority can grant up to `allocation` usernames. To top up their allocation, they
+		/// should just issue (or request via governance) a new `add_username_authority` call.
+		#[pallet::call_index(15)]
+		#[pallet::weight(T::WeightInfo::add_username_authority())]
+		pub fn add_username_authority(
+			origin: OriginFor<T>,
+			authority: AccountIdLookupOf<T>,
+			suffix: Vec<u8>,
+			allocation: u32,
+		) -> DispatchResult {
+			T::UsernameAuthorityOrigin::ensure_origin(origin)?;
+			let authority = T::Lookup::lookup(authority)?;
+			Self::validate_username(&suffix).map_err(|_| Error::<T>::InvalidSuffix)?;
+			let suffix = Suffix::try_from(suffix).map_err(|_| Error::<T>::InvalidSuffix)?;
+			// The authority may already exist, but we don't need to check. They might be changing
+			// their suffix or adding allocation, so we just want to overwrite whatever was there.
+			UsernameAuthorities::<T>::insert(
+				&authority,
+				AuthorityProperties { suffix, allocation },
+			);
+			Self::deposit_event(Event::AuthorityAdded { authority });
+			Ok(())
+		}
+
+		/// Remove `authority` from the username authorities.
+		#[pallet::call_index(16)]
+		#[pallet::weight(T::WeightInfo::remove_username_authority())]
+		pub fn remove_username_authority(
+			origin: OriginFor<T>,
+			authority: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			T::UsernameAuthorityOrigin::ensure_origin(origin)?;
+			let authority = T::Lookup::lookup(authority)?;
+			UsernameAuthorities::<T>::remove(&authority);
+			Self::deposit_event(Event::AuthorityRemoved { authority });
```

Comment:
If the authority didn't exist before, this event may be misleading. Unless line 1003 returns early in that case.